src/main/java/bio/terra/common/tracing/TracingHook.java

```diff
@@ -0,0 +1,214 @@
+package bio.terra.common.tracing;
+
+import bio.terra.stairway.*;
+import io.opencensus.common.Scope;
+import io.opencensus.trace.*;
+import io.opencensus.trace.propagation.SpanContextParseException;
+import java.util.Base64;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link StairwayHook} to add support for tracing execution of Stairway flights with OpenCensus.
+ *
+ * <p>This hook uses Spans at 3 different levels. From lowest to highest:
+ * <li>A Span around the execution of each Step, one Span per Step execution. All of the {@link
+ *     Step#doStep(FlightContext)} {@link Step#undoStep(FlightContext)} will be executed in the
+ *     context of this Span.
+ * <li>A Span around the execution of each Flight, one Span per {@link Flight#run()}. Note that a
```

Comment:
This part was a little conceptually jarring to me.

The most ideal output, from a naive developer's perspective, would be for there to be one top-level span for the entire flight progression, regardless of pause and resume events. Identifying pause / resume events with https://opencensus.io/tracing/span/time_events/annotation/ would be icing on the cake.

Without going too overboard, I wonder if there's some way to handle this. Maybe you could store the SpanContext for the top-level Flight span within the working map, and do a get-or-create in the initial startFlight call?

To some extent, this would feel like what's already being done in the case where a manual submission context span needs to be created. Maybe there's ultimately not a need for both? i.e. right now we conditionally create a Submission span to act as the parent of N flight-execution spans, but maybe we can always create one Flight span (optionally linked to a service-provided submission span) to act as the parent of N slight-step spans.

Reply:
Nope! To have a single span for the entire flight progression, we would have to be able to serialize and resume that span in case it hops between Stairway instances. This is explicitly not possible through the public APIs of the OpenCensus library i.e. you cannot create a Span from its own SpanContext. You can only create a new Span. See the methods on https://javadoc.io/static/io.opencensus/opencensus-api/0.28.3/io/opencensus/trace/Tracer.html and note that their are no public constructors of SpanBuilder or Span.
A Span is local to a process. We can propagate SpanContexts to create child Spans (when the child *will not* outlive the parent Span) or create Spans with a link to the parent Span (when the child *may* outlive the parent Span.)

Reply:
Thanks for the explanation – I think I understand now!

Thinking from a user's perspective on the Trace visualization page, it would still be really nice to know whether a given Flight span represents the complete vs. partial execution of a given flight. To whatever extent we can surface that information from within a hook (or bubble up some feature requests to Stairway to propagate that context), that would be nice... but obviously not mission critical, especially if this context is at least captured in the correlated logs.

And thinking about potential future analytical use cases (e.g. digging into span data via bigquery), getting the overall runtime of a flight won't be as simple as looking at the duration of a span w/ that classname. You'd have to do something like min/max timestamp of all spans with the same classname *and* the same flight ID. Or, otherwise perhaps exclude spans that don't encapsulate an entire flight, since we expect that to be the minority... though this is only possible if we have some annotations that let us filter out these spans.